cpp/src/io/utilities/type_inference.cuh

```diff
@@ -0,0 +1,261 @@
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_scalar.hpp>
+
+#include <thrust/distance.h>
+#include <thrust/tuple.h>
+
+#include <cstddef>
+
+namespace cudf {
+namespace io {
+namespace detail {
+/**
+ * @brief Structure for type inference options
+ */
+struct inference_options_view {
+  cudf::detail::trie_view trie_true;
+  cudf::detail::trie_view trie_false;
+  cudf::detail::trie_view trie_na;
+  char quote_char;
+};
+
+struct inference_options {
+  cudf::detail::optional_trie trie_true;
+  cudf::detail::optional_trie trie_false;
+  cudf::detail::optional_trie trie_na;
+  char quote_char = '"';
+
+  [[nodiscard]] inference_options_view view() const
+  {
+    return {cudf::detail::make_trie_view(trie_true),
+            cudf::detail::make_trie_view(trie_false),
+            cudf::detail::make_trie_view(trie_na),
+            quote_char};
+  }
+};
+
+/**
+ * @brief Returns true is the input character is a valid digit.
+ * Supports both decimal and hexadecimal digits (uppercase and lowercase).
+ *
+ * @param c Character to check
+ * @param is_hex Whether to check as a hexadecimal
+ *
+ * @return `true` if it is digit-like, `false` otherwise
+ */
+__device__ __inline__ bool is_digit(char const c, bool const is_hex = false)
+{
+  if (c >= '0' && c <= '9') return true;
+
+  if (is_hex) {
+    if (c >= 'A' && c <= 'F') return true;
+    if (c >= 'a' && c <= 'f') return true;
+  }
+
+  return false;
+}
+
+/**
+ * @brief Returns true if the counters indicate a potentially valid float.
+ * False positives are possible because positions are not taken into account.
+ * For example, field "e.123-" would match the pattern.
+ */
+__device__ __inline__ bool is_like_float(
+  long len, long digit_cnt, long decimal_cnt, long dash_cnt, long exponent_cnt)
+{
+  // Can't have more than one exponent and one decimal point
+  if (decimal_cnt > 1) return false;
+  if (exponent_cnt > 1) return false;
+  // Without the exponent or a decimal point, this is an integer, not a float
+  if (decimal_cnt == 0 && exponent_cnt == 0) return false;
+
+  // Can only have one '-' per component
+  if (dash_cnt > 1 + exponent_cnt) return false;
+
+  // If anything other than these characters is present, it's not a float
+  if (digit_cnt + decimal_cnt + dash_cnt + exponent_cnt != len) return false;
+
+  // Needs at least 1 digit, 2 if exponent is present
+  if (digit_cnt < 1 + exponent_cnt) return false;
+
+  return true;
+}
+
+template <typename ColumnStringIter>
+__global__ void detect_column_type_kernel(inference_options_view const options,
+                                          device_span<char const> const data,
+                                          ColumnStringIter column_strings_begin,
+                                          std::size_t const size,
+                                          cudf::io::column_type_histogram* column_info)
+{
+  for (auto idx = threadIdx.x + blockDim.x * blockIdx.x; idx < size;
+       idx += gridDim.x + blockDim.x) {
```

Comment:
I recognized this has propagated from the original while loop to my suggestion. This will make like a 100x difference in performance. Easy win 😀 
```suggestion
  for (auto idx = threadIdx.x + blockDim.x * blockIdx.x; idx < size;
       idx += gridDim.x * blockDim.x) {
```

Reply:
Thanks for fixing this so quickly, @PointKernel. With this one in, performance improvements (e.g., accumulating using shared memory atomics) aren't as much of a priority anymore.